sympy/matrices/determinant.py

```diff
@@ -398,21 +404,21 @@ def _charpoly(M, x='lambda', simplify=_simplify):
 
     det
     """
-
-    if not M.is_square:
-        raise NonSquareMatrixError()
-    if M.is_lower or M.is_upper:
-        diagonal_elements = M.diagonal()
-        x = uniquely_named_symbol(x, diagonal_elements, modify=lambda s: '_' + s)
-        m = 1
-        for i in diagonal_elements:
-            m = m * (x - simplify(i))
-        return PurePoly(m, x)
-
-    berk_vector = _berkowitz_vector(M)
-    x = uniquely_named_symbol(x, berk_vector, modify=lambda s: '_' + s)
-
-    return PurePoly([simplify(a) for a in berk_vector], x)
+    from sympy.polys.domains.pythonrational import PythonRational
+    from sympy.polys.domains.gaussiandomains import GaussianRational
+    from sympy import EX
+    m = 0
+    coeff = _charpoly_DOM(M)
+    l = len(coeff)
+    x = uniquely_named_symbol(x, modify=lambda s: '_' + s)
+    for i in range(l):
+        if not isinstance(coeff[i], PythonRational):
+            if isinstance(coeff[i], GaussianRational):
+                coeff[i] = EX.to_sympy(coeff[i])
+            else:
+                coeff[i] = EX.to_sympy(coeff[i])
+        m = m + coeff[i]*x**(l-i-1)
+    return PurePoly(m, x)
```

Comment:
I suggest reading this carefully:
https://docs.sympy.org/dev/modules/polys/domainsintro.html

The way to convert a domain element to Expr is using the `Domain.to_sympy` method but it should be the domain for that particular domain element. Here you are using the `EX` domain but the elements might not be of the `EX` domain.

Also you are converting from domain element to Expr only to then convert back to the domain in the call to PurePoly.

The way to create a Poly directly from a list of domain elements is like:
```python
In [41]: K = ZZ[x]

In [42]: coeffs = [K.from_sympy(e) for e in [x, 1+x, 1+x**2]]

In [43]: coeffs
Out[43]: [x, x + 1, x**2 + 1]

In [44]: Poly(coeffs, z, domain=K)
Out[44]: Poly(x*z**2 + (x + 1)*z + x**2 + 1, z, domain='ZZ[x]')
```
It would be better to do that in the `_charpoly_DOM` function though.

Reply:
Thank you for your assistance.
I went through the doc, this commit seems to pass all the test cases for `test_determinant`.
@oscarbenjamin, Kindly review the latest commit.
Thanks.

Reply:
The approach now is overcomplicating things. There is no need to convert the coefficients back to Expr. They can be passed directly to Poly as I showed above.

Reply:
@oscarbenjamin, this implementation doesn't work for the case
```
n = DeterminantOnlyMatrix(3, 3, [x, 0, 0, a, y, 0, b, c, z])
assert n.charpoly() == Poly(t**3 - (x+y+z)*t**2 + t*(x*y+y*z+x*z) - x*y*z , t, domain = 'QQ')
```
Otherwise, it seems to pass all test cases for `test_determinant`.
Moreover, we need to add the `domain = QQ`.
```
def _charpoly_DOM(M, x):
    DOM = DomainMatrix.from_Matrix(M, field=True, extension=True)
    K = DOM.domain
    coeff = DOM.charpoly()
    coeffs = []
    flag = True
    for e in coeff:
        if isinstance(e, PythonRational) or isinstance(e, GaussianRational):
            coeffs.append(e)
            flag = False
        else:
            coeffs.append(K.from_sympy(e))
    if flag:
        x = uniquely_named_symbol(x, coeffs, modify=lambda s: '_' + s)
    else:
        if x == 'lambda':
            x = Symbol('x')
        else:
            x = uniquely_named_symbol(x, modify=lambda s: '_' + s)
    return Poly(coeffs, x, domain = DOM.domain)
```
The traceback: 
```
Traceback (most recent call last):
  File "c:\users\harshit yadav\sympy\sympy\matrices\tests\test_determinant.py", line 387, in test_charpoly
    assert n.charpoly() == Poly(t**3 - (x+y+z)*t**2 + t*(x*y+y*z+x*z) - x*y*z , t, domain = 'QQ')
  File "c:\users\harshit yadav\sympy\sympy\matrices\matrices.py", line 117, in charpoly
    return _charpoly(self, x=x, simplify=simplify)
  File "c:\users\harshit yadav\sympy\sympy\matrices\determinant.py", line 426, in _charpoly
    return _charpoly_DOM(M, x)
  File "c:\users\harshit yadav\sympy\sympy\matrices\determinant.py", line 344, in _charpoly_DOM
    coeffs.append(K.from_sympy(e))
  File "c:\users\harshit yadav\sympy\sympy\polys\domains\fractionfield.py", line 77, in from_sympy
    return self.field.from_expr(a)
  File "c:\users\harshit yadav\sympy\sympy\polys\fields.py", line 271, in from_expr
    frac = self._rebuild_expr(expr, mapping)
  File "c:\users\harshit yadav\sympy\sympy\polys\fields.py", line 265, in _rebuild_expr
    return _rebuild(sympify(expr))
  File "c:\users\harshit yadav\sympy\sympy\core\sympify.py", line 355, in sympify
    raise SympifyError(a)
sympy.core.sympify.SympifyError: SympifyError: 1
```
So this error is arising as here the `coeffs` are of type `FracElement`, and `sympify` seems to raise error for it.
Kindly guide how to fix it, is this a bug, or I am missing something in the implementation.
Thanks.

Reply:
The `K.from_sympy` method is to convert from Expr to a domain element. You are passing in a domain element which is backwards.

You can just use the original matrix `M` for `uniquely_named_symbol` anyway.

Reply:
Sure, it was very immature of me.
Thank you.